ArchiSteamFarm.Tests/Bot.cs

```diff
@@ -0,0 +1,197 @@
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void MultipleSets() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(1),
+				CreateCard(2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 2, amountPerCard: 2);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 2 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 2 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void MultipleSetsDifferentAmount() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, 2),
+				CreateCard(2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 2, amountPerCard: 2);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 2 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 2 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void MoreCardsThanNeeded() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(1),
+				CreateCard(2),
+				CreateCard(3)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 3, amountPerCard: 1);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 1 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 1 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 3), 1 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void CardsWithOtherRarity() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, rarity: Steam.Asset.ERarity.Common),
+				CreateCard(1, rarity: Steam.Asset.ERarity.Rare)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 1, amountPerCard: 1);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 1 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void CardsWithOtherType() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, type: Steam.Asset.EType.TradingCard),
+				CreateCard(1, type: Steam.Asset.EType.FoilTradingCard)
```

Comment:
And this should have exactly the same tests as the one with rarity (so also 3 total). In addition to that there should be at least 2-3 tests doing the same logic but with different combinations of type and rarity.